test-cypress/cypress/commands/api/visits.ts

```diff
@@ -5,39 +5,117 @@ import { v1ApiPath, getCreds } from "../server";
+      {
+        user,
+        camera,
+        expectedVisits
+      }
+    );
 
-    if (camera) {
-      where.DeviceId = getCreds(camera).id;
-    }
+    checkVisitsMatch(user, camera, expectedVisits);
+  }
+);
 
-    const params = {
-      where: JSON.stringify(where),
-      limit: 100
-    };
-
-    const visitsNum = noVists;
-    cy.request({
-      method: "GET",
-      url: v1ApiPath("recordings/visits", params),
-      headers: getCreds(user).headers
-    }).then((response) => {
-      expect(response.body.numVisits).to.eq(visitsNum);
+Cypress.Commands.add(
+  "checkVisits",
+  (user: string, camera: string, expectedVisits: ComparableVisit[]) => {
+    logTestDescription(`Check visits match ${JSON.stringify(expectedVisits)}`, {
+      user,
+      camera,
+      expectedVisits
     });
+
+    checkVisitsMatch(user, camera, expectedVisits);
   }
 );
 
+function checkVisitsMatch(
+  user: string,
+  camera: string,
+  expectedVisits: ComparableVisit[]
+) {
+  const where: VisitsWhere = {
+    duration: { $gte: "0" },
+    type: "thermalRaw"
+  };
+
+  if (camera) {
+    where.DeviceId = getCreds(camera).id;
+  }
+
+  const params = {
+    where: JSON.stringify(where),
+    limit: 100
+  };
+
+  cy.request({
+    method: "GET",
+    url: v1ApiPath("recordings/visits", params),
+    headers: getCreds(user).headers
+  }).then((response) => {
+    checkResponseMatches(response, expectedVisits);
+  });
+}
+
+function checkResponseMatches(
+  response: any,
+  expectedVisits: ComparableVisit[]
+) {
+  const responseVisits = response.body.visits;
+
+  expect(
+    responseVisits.length,
+    `Number of visits to be ${responseVisits.length}`
+  ).to.eq(expectedVisits.length);
+  const increasingDateResponseVisits = responseVisits.reverse();
+
+  // pull out the bits we care about
+  const responseVisitsToCompare: ComparableVisit[] = [];
+  for (var i = 0; i < expectedVisits.length; i++) {
+    const expectedVisit = expectedVisits[i];
+    const completeResponseVisit = increasingDateResponseVisits[i];
+    const simplifiedResponseVisit: ComparableVisit = {};
+
+    if (expectedVisit.tag) {
+      simplifiedResponseVisit.tag = completeResponseVisit.what || "<null>";
+    }
+
+    if (expectedVisit.recordings) {
+      const recordingIds = completeResponseVisit.events.map((ev) => {
+        return ev.recID;
+      });
+      const uniqueRecordingIds = recordingIds.reduce((acc, recId) => {
+        if (!acc.includes(recId)) {
+          acc.push(recId);
+        }
+        return acc;
+      }, []);
+      simplifiedResponseVisit.recordings = uniqueRecordingIds.length;
+    }
+
+    if (expectedVisit.start) {
+      simplifiedResponseVisit.start = completeResponseVisit.start;
+    }
+
+    if (expectedVisit.end) {
+      simplifiedResponseVisit.end = completeResponseVisit.end;
+    }
+
+    responseVisitsToCompare.push(simplifiedResponseVisit);
+  }
+  expect(JSON.stringify(responseVisitsToCompare)).to.eq(
```

Comment:
I'm not 100% sure if we can rely on two different objects that are the same having the same JSON.stringify representation - there seems to be variance across JS engines as to whether stringify is actually deterministic in terms of things like object key order.  It sounds like object key order might actually be dependent on the creation order of the object properties?  If you do 
```
let obj1 = {};
let obj2 = {};
obj1["a"] = true;
obj1["b"] = true;
obj2["b"] = true;
obj2["a"] = true;
```
and stringify both, the strings will not be the same.  This might cause things to fail sometimes?

Reply:
Yes.   These are tests so if we see a problem then we'll do it differently.  If the are arrays are different it will fail and that what's we need to happen.